pandas/core/algorithms.py

```diff
@@ -45,40 +37,192 @@
-        values = PeriodIndex(values).asi8
-    elif is_categorical_dtype(values):
-        values = values.values.codes
-    elif isinstance(values, (ABCSeries, ABCIndex)):
-        values = values.values
-
-    if is_signed_integer_dtype(values):
+    if (needs_i8_conversion(values) or
+            is_period_dtype(dtype) or
+            is_datetime64_any_dtype(dtype) or
+            is_timedelta64_dtype(dtype)):
+        if is_period_dtype(values) or is_period_dtype(dtype):
+            from pandas import PeriodIndex
+            values = PeriodIndex(values)
+            dtype = values.dtype
+        elif is_timedelta64_dtype(values) or is_timedelta64_dtype(dtype):
+            from pandas import TimedeltaIndex
+            values = TimedeltaIndex(values)
+            dtype = values.dtype
+        elif is_datetimetz(values) or is_datetime64tz_dtype(dtype):
+            from pandas import DatetimeIndex
+            values = DatetimeIndex(values)
+            dtype = values.dtype
+        else:
+            from pandas import DatetimeIndex
+            values = DatetimeIndex(values)
+            dtype = values.dtype
+
+        return values.asi8.view('int64'), dtype, 'int64'
+
+    elif is_categorical_dtype(values) or is_categorical_dtype(dtype):
+        values = getattr(values, 'values', values)
+        values = values.codes
+        dtype = 'category'
+
+        # we are actually coercing to int64
+        # until our algos suppport int* directly (not all do)
         values = _ensure_int64(values)
-    elif is_unsigned_integer_dtype(values):
-        values = _ensure_uint64(values)
-    elif is_complex_dtype(values):
-        values = _ensure_float64(values)
-    elif is_float_dtype(values):
-        values = _ensure_float64(values)
-    else:
+
+        return values, dtype, 'int64'
+
+    values = np.asarray(values)
+
+    try:
+        if is_bool_dtype(values) or is_bool_dtype(dtype):
+            # we are actually coercing to uint64
+            # until our algos suppport uint8 directly (see TODO)
+            values = values.view('uint64')
+            dtype = 'bool'
+            ndtype = 'uint64'
+        elif is_signed_integer_dtype(values) or is_signed_integer_dtype(dtype):
+            values = _ensure_int64(values)
+            ndtype = dtype = 'int64'
+        elif (is_unsigned_integer_dtype(values) or
+              is_unsigned_integer_dtype(dtype)):
+            values = _ensure_uint64(values)
+            ndtype = dtype = 'uint64'
+        elif is_complex_dtype(values) or is_complex_dtype(dtype):
+            values = _ensure_float64(values)
+            ndtype = dtype = 'float64'
+        elif is_float_dtype(values) or is_float_dtype(dtype):
+            values = _ensure_float64(values)
+            ndtype = dtype = 'float64'
+        else:
+            values = _ensure_object(values)
+            ndtype = dtype = 'object'
+
+    except (TypeError, ValueError):
+        # object array conversion will fail
```

Comment:
not clear what would actually hit this? should't object fall to the `else` branch above?